packages/react-dom/src/server/ReactPartialRendererHooks.js

```diff
@@ -0,0 +1,374 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+import type {ReactContext} from 'shared/ReactTypes';
+
+import invariant from 'shared/invariant';
+import warning from 'shared/warning';
+
+type BasicStateAction<S> = S | (S => S);
```

Comment:
This is more sound I think. In current case flow will always refine any value and function params and return value won't be checked. I had the same problem in powerplug.
```
type BasicStateAction<S> = (S => S) | S;
```

Reply:
Yeah we found the same thing when testing at FB. Forgot to update in the upstream brach. :)